README.md
# CloudVAMP - Cloud Virtual Machine Automatic Memory Procurement

## ABOUT

### What is CloudVAMP?

CloudVAMP stands for "Cloud Virtual Machine Automatic Memory Procurement" and it is an automatic system that enables and manages memory oversubscription in a Cloud on-premises platform based in OpenNebula.

### Why CloudVAMP?

Users typically deploy Virtual Machines (VMs) with more resources allocated (e.g. memory) than actually needed by the applications running in the VMs. This might happen for different reasons: i) Users are unaware of the actual requirements of applications due to a lack of profiling; ii) The applications might show dynamic memory consumption during their execution; iii) The Cloud Management Framework (e.g. OpenNebula, OpenStack) determines the amount of memory allocated for the VM out of a set of predefined templates.

### How does CloudVAMP work?

CloudVAMP steals the memory that is not used in the running VMs on an OpenNebula Cloud, and makes it available for other VMs. If the memory is later needed by the original VM, it is returned to it.

### What are the technical details?

If an excess of free memory for a VM is detected, CloudVAMP changes the amount of memory allocated to the VM via the underlying hypervisor (i.e. KVM) so that it fits the used memory (plus an additional margin of memory), thus considering the running applications. The host that is hosting the VM will then have an extra amount of "stolen memory" that can be used for the deployment of additional VMs on that host. Therefore, CloudVAMP introduces the ability of oversubscription for an on-premises Cloud.

CloudVAMP also prevents overloading, i.e. when the sum of memory allocated to the VMs within a host is greater than the physical amount of memory available for the host. For this, if additional memory is reclaimed by a VM, CloudVAMP uses live migration of VMs across hosts in order to safely prevent overloading without VM downtime.

## Installation

The current version of CloudVAMP is available for OpenNebula. It has been tested in version 4.8 but it is likely to work with newer versions. The installation consists of three steps:

1. Cloud Vertical Elasticity Manager (CVEM), installed in the OpenNebula front-end host.
2. Memory Reporter (MR), installed in the VM
3. Memory Oversubscription Granter (MOG), installed in the OpenNebula front-end host.

### Getting CloudVAMP from git

#### Download CloudVAMP from git:
  ```
    $ cd /tmp
    $ git clone https://github.com/grycap/cloudvamp
  ```
or
  ```
    $ cd /tmp
    $ wget https://github.com/grycap/cloudvamp/archive/master.zip
    $ unzip cloudvamp-master.zip
    $ mv cloudvamp-master cloudvamp
  ```

### Installing CVEM

#### Requirements

CVEM uses the library [cpyutils](https://github.com/grycap/cpyutils). Follow the instructions there to install it.

#### Install in the system path

Enter the CVEM directory in the downloaded files and perform the setup installation.
  ```
    $ cd /tmp/cloudvamp/cvem
    $ python setup install
  ```

#### Install in a specific path

Select a proper path where to install the CVEM service (i.e. `/usr/local/cvem`, `/opt/cvem` or other). This path will be called `CVEM_PATH`.
  ```
    $ cp -r cd /tmp/cloudvamp/cvem /usr/local
  ```

Finally you must copy (or link) `$CVEM_PATH/scripts/cvem` file to `/etc/init.d` directory.
  ```
    $ ln -s /usr/local/cvem/scripts/cvem /etc/init.d
  ```

#### Configuration

If you want the CVEM service to be automatically started at boot time, execute the following commands:

On RedHat Systems:
  ```
    $ chkconfig cvem on
  ```
On Debian Systems:
  ```
    $ update-rc.d cvem start 99 2 3 4 5 . stop 05 0 1 6 .
  ```
Or you can do it manually:
  ```
    $ ln -s /etc/init.d/cvem /etc/rc2.d/S99cvem
    $ ln -s /etc/init.d/cvem /etc/rc3.d/S99cvem
    $ ln -s /etc/init.d/cvem /etc/rc5.d/S99cvem
    $ ln -s /etc/init.d/cvem /etc/rc1.d/K05cvem
    $ ln -s /etc/init.d/cvem /etc/rc6.d/K05cvem
  ```
Adjust the installation path by setting the DAEMON variable at `/etc/init.d/cvem` to the path where the CVEM `cvemd.py` file is installed (e.g. `/usr/local/cvem/cvemd.py`), or set the name of the script file (`cvemd.py`) if the file is in the environment variable `PATH`.

Finally edit the configurations files *.cfg located in the etc/ directory where the CVEM is installed (e.g `/opt/cvem/etc`) or in the `/etc/cvem` directory:

* `cvem.cfg`: CVEM configuration file.
* `one.cfg`: OpenNebula specific configuration file.

### Installing the Memory Reporter (MR)

The memory reporter relies on the OpenNebula contextualization scripts. Therefore, it must be installed on the Virtual Machine images used in the platform. See more information in the [OpenNebula Documentation](http://docs.opennebula.org/4.12/user/virtual_machine_setup/bcont.html).

It uses the OneGate system to publish and get the monitored information so OneGate must be installed and configured, as indicated in the [OpenNebula's OneGate Documentation](http://archives.opennebula.org/documentation:rel4.4:onegate_usage).

The first step is to upload the contextualization scripts (`onegate_init.sh`, `onegate_publisher.sh`) to the OpenNebula "Files & Kernels" section. If you are using OpenNebula versions 4.8 or newer the files are located in `/tmp/cloudvamp/onegate_scripts/` however if you are using versions 4.6 or older the files are located in `/tmp/cloudvamp/onegate_scripts/old`   

Then all the templates of the VMs must be configured to activate the OpenNebula contextualization Token, include both files, and set `onegate_init.sh` as an init script:

  ```
    CONTEXT = [
        FILES_DS = "$FILE[IMAGE_ID=<init_file_id>] $FILE[IMAGE_ID=<publisher_file_id>]",
        INIT_SCRIPTS = "onegate_init.sh",
        TOKEN = "YES" ]
  ```

### Installing the Memory Oversubscription Granter (MOG)

#### Creating the IM and the VMM

The MOG, installed in the OpenNebula front-end node, consists of two parts: an Information Manager (IM) and a Virtual Machine Manager (VMM), which are concepts from OpenNebula. The code has been tested in Ubuntu, with a ONE 4.8 installation from the official OpenNebula repositories.

Create a folder for the CloudVAMP VMM at ONE's vmm folder (usually `/var/lib/one/remotes/vmm`) and create links to the files in the kvm folder:

  ```
    $ mkdir /var/lib/one/remotes/vmm/cloudvamp
    $ cd /var/lib/one/remotes/vmm/cloudvamp
  ```

Now copy the new files in the folder

  ```
    $ cp /tmp/cloudvamp/opennebula/vmm/cloudvamp/* /var/lib/one/remotes/vmm/cloudvamp/
  ```

Finally link the other file in the folder (please ignore the messages that state that the files already exist):
  ```
    $ ln -s ../kvm/* /var/lib/one/remotes/vmm/cloudvamp/
  ```

Create a folder for cloudvamp im probes files at ONE's im folder (usually `/var/lib/one/remotes/im`) and create links to the files in kvm folder:

  ```
    $ mkdir /var/lib/one/remotes/im/cloudvamp-probes.d
  ```

Now copy the new files in the folder

  ```
    $ cp /tmp/cloudvamp/opennebula/im/cloudvamp-probes.d/* /var/lib/one/remotes/im/cloudvamp-probes.d/
  ```

Link the files to the kvm probes and set the values for the cloudvamp probes

  ```
    $ ln -s /var/lib/one/remotes/im/kvm-probes.d/* /var/lib/one/remotes/im/cloudvamp-probes.d/
    $ cd /var/lib/one/remotes/im/cloudvamp-probes.d
    $ rm kvm.rb
    $ chmod 644 poll.sh
  ```

* NOTE: take into account that the file "poll.sh" must not have permission for execution (i.e. its permission mask is 0644).

Finally, create a folder for the cloudvamp im as a link to the kvm one:

  ```
    $ ln -s /var/lib/one/remotes/im/kvm.d /var/lib/one/remotes/im/cloudvamp.d
  ```

#### Activating CloudVAMP in ONE

Edit the one configuration file (i.e. `/etc/one/oned.conf`) and add the following lines

  ```
    IM_MAD = [
        name       = "cloudvamp",
        executable = "one_im_ssh",
        arguments  = "-r 3 -t 15 cloudvamp" ]


    VM_MAD = [
      name       = "cloudvamp",
      executable = "one_vmm_exec",
      arguments  = "-t 15 -r 0 cloudvamp",
      default    = "vmm_exec/vmm_exec_kvm.conf",
      type       = "kvm" ]
  ```

And restart ONE
  ```
  $ su - oneadmin
  $ one restart
  ```

#### Activating CloudVAMP for the nodes

Now you should delete (or deactivate) the nodes under the control of ONE and create them again using the cloudvamp im and vmm. This will only apply to these VMs that are running the KVM hypervisor

> In this particular case we are using dummy network

  ```
  $ onehost disable myhost01
  $ onehost create myhost01 -i cloudvamp -v cloudvamp -n dummy
  ```

#### Tuning CloudVAMP memory management

In case that you do not want that all the memory stolen from the running VMs is dedicated to other VMs, you can tune the $O value from file `/var/lib/one/remotes/im/cloudvamp-probes.d/cloudvamp.rb`.

> E.g. If you want that only the 80% of borrowed memory is dedicated for possible oversubscription, please search for the line $O = 1.0 and set it to $O = 0.8 (take into account that it is a floating point number, in case that you get back to 100%)

## Copyright and License
Code and documentation (c) 2015 - Universitat Politecnica de Valencia - GRyCAP. Licensed under the Apache License 2.0

## Publication
CloudVAMP additional details can be found in the following academic publication:

> Germán Moltó, Miguel Caballer, Carlos de Alfonso, "Automatic Memory-based Vertical Elasticity and Oversubscription on Cloud Platform", to appear in Future Generation Computer Systems (The International Journal of eScience).


cvem/changelog
CVEM 0.10

    * Initial version

CVEM 0.20

    * Add CLUES support

CVEM 0.21

    * Bugfix of not finding logging.conf file
    * Enable to specify MEM_FREE_MIN and MEM_OVER per VM
    * Bugfix with exponential backoff
	* Code refactoring

cvem/connectors/one/OpenNebula.py
# -------------------------------------------------------------------------- #
# Copyright 2015, Universitat Politecnica de Valencia                        #
#                                                                            #
# Licensed under the Apache License, Version 2.0 (the "License"); you may    #
# not use this file except in compliance with the License. You may obtain    #
# a copy of the License at                                                   #
#                                                                            #
# http://www.apache.org/licenses/LICENSE-2.0                                 #
#                                                                            #
# Unless required by applicable law or agreed to in writing, software        #
# distributed under the License is distributed on an "AS IS" BASIS,          #
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.   #
# See the License for the specific language governing permissions and        #
# limitations under the License.                                             #
#--------------------------------------------------------------------------- #

from cpyutils.xmlobject import XMLObject
from cpyutils.timeoutxmlrpccli import ServerProxy
from cvem.config import logger, Config
from cvem.CMPInfo import VirtualMachineInfo, HostInfo, CMPInfo
from cvem.Monitor import Monitor
from config_one import ConfigONE

# classes to parse the results of the ONE API using xmlobject
class NIC(XMLObject):
		values = ['BRIDGE', 'IP', 'MAC', 'NETWORK', 'VNID']

class OS(XMLObject):
		values = ['BOOT', 'ROOT']

class GRAPHICS(XMLObject):
		values = ['LISTEN', 'TYPE']

class DISK(XMLObject):
		values = ['CLONE','READONLY','SAVE','SOURCE','TARGET' ]

class TEMPLATE(XMLObject):
		# In ONE versions <= 4.12 PACKEDMEMORY, MAXMEMORY and REALMEMORY are located here
		values = [ 'CPU', 'MEMORY', 'NAME', 'RANK', 'REQUIREMENTS', 'VMID', 'VCPU', 'PACKEDMEMORY', 'MAXMEMORY', 'REALMEMORY' ]
		tuples = { 'GRAPHICS': GRAPHICS, 'OS': OS }
		tuples_lists = { 'DISK': DISK, 'NIC': NIC }
		numeric = [ 'CPU', 'MEMORY', 'VCPU' ]
		noneval = 0

class HISTORY(XMLObject):
		values = ['SEQ', 'HOSTNAME', 'HID', 'STIME', 'ETIME', 'PSTIME', 'PETIME', 'RSTIME', 'RETIME' ,'ESTIME', 'EETIME', 'REASON' ]

class HISTORY_RECORDS(XMLObject):
	tuples_lists = { 'HISTORY': HISTORY }

class USER_TEMPLATE(XMLObject):
		values = [ 'MEM_FREE', 'MEM_TOTAL', 'MEM_TOTAL_REAL', 'MIN_FREE_MEM', 'MEM_OVER']
		numeric = [ 'MEM_FREE', 'MEM_TOTAL', 'MEM_TOTAL_REAL', 'MIN_FREE_MEM', 'MEM_OVER']

class VM(XMLObject):
		STATE_INIT=0
		STATE_PENDING=1
		STATE_HOLD=2
		STATE_ACTIVE=3
		STATE_STOPPED=4
		STATE_SUSPENDED=5
		STATE_DONE=6
		STATE_FAILED=7
		STATE_STR = {'0': 'init', '1': 'pending', '2': 'hold', '3': 'active', '4': 'stopped', '5': 'suspended', '6': 'done', '7': 'failed' }
		LCM_STATE_STR={'0':'init','1':'prologing','2':'booting','3':'running','4':'migrating','5':'saving (stop)','6':'saving (suspend)','7':'saving (migrate)', '8':'prologing (migration)', '9':'prologing (resume)', '10': 'epilog (stop)','11':'epilog', '12':'cancel','13':'failure','14':'delete','15':'unknown'}
		# In ONE versions >= 4.14 PACKEDMEMORY, MAXMEMORY and REALMEMORY are located here
		values = [ 'ID','UID','NAME','LAST_POLL','STATE','LCM_STATE','DEPLOY_ID','MEMORY','CPU','NET_TX','NET_RX', 'STIME','ETIME', 'PACKEDMEMORY', 'MAXMEMORY', 'REALMEMORY' ]
		tuples = { 'TEMPLATE': TEMPLATE, 'HISTORY_RECORDS': HISTORY_RECORDS, 'USER_TEMPLATE': USER_TEMPLATE }
		numeric = [ 'ID', 'UID', 'STATE', 'LCM_STATE', 'STIME','ETIME' ]

class VM_POOL(XMLObject):
	tuples_lists = { 'VM': VM }

class LEASE(XMLObject):
	values = [ 'IP', 'MAC', 'USED' ]

class TEMPLATE_VNET(XMLObject):
	values = [ 'BRIDGE', 'NAME', 'TYPE', 'NETWORK_ADDRESS' ]
	tuples_lists = { 'LEASES': LEASE }

class LEASES(XMLObject):
	tuples_lists = { 'LEASE': LEASE }
	
class RANGE(XMLObject):
	values = [ 'IP_START', 'IP_END' ]
	
class AR(XMLObject):
	values = [ 'IP', 'MAC', 'TYPE', 'ALLOCATED', 'GLOBAL_PREFIX', 'AR_ID' ]
	
class AR_POOL(XMLObject):
	tuples_lists = { 'AR': AR }

class VNET(XMLObject):
	values = [ 'ID', 'UID', 'GID', 'UNAME', 'GNAME', 'NAME', 'TYPE', 'BRIDGE', 'PUBLIC' ]
	tuples = { 'TEMPLATE': TEMPLATE_VNET, 'LEASES': LEASES, 'RANGE': RANGE, 'AR_POOL':AR_POOL }
	
class VNET_POOL(XMLObject):
	tuples_lists = { 'VNET': VNET }

class HOST_SHARE(XMLObject):
	values = [ 'MEM_USAGE', 'MAX_MEM', 'FREE_MEM', 'FREE_CPU', 'MAX_CPU' ]
	numeric = [ 'MEM_USAGE', 'MAX_MEM', 'FREE_MEM', 'FREE_CPU', 'MAX_CPU' ]

class TEMPLATE_HOST(XMLObject):
	values = [ 'PACKEDMEMORY' ]
	numeric = [ 'PACKEDMEMORY' ]

class VMS(XMLObject):
	values_lists = { 'ID' }

class HOST(XMLObject):
	STATE_INIT=0
	STATE_MONITORING_MONITORED=1
	STATE_MONITORED=2
	STATE_ERROR=3
	STATE_DISABLED=4
	STATE_MONITORING_ERROR=5
	STATE_MONITORING_INIT=6
	STATE_MONITORING_DISABLED=7
	INVALID_STATES = [STATE_ERROR, STATE_DISABLED, STATE_MONITORING_ERROR, STATE_MONITORING_DISABLED]
	values = [ 'ID', 'LAST_MON_TIME', 'NAME', 'STATE' ]
	numeric = [ 'ID', 'STATE' ]
	tuples = { 'HOST_SHARE': HOST_SHARE, 'TEMPLATE': TEMPLATE_HOST, 'VMS': VMS}
	
class HOST_POOL(XMLObject):
	tuples_lists = { 'HOST': HOST }
	
	
class OpenNebula(CMPInfo):
	"""
	OpenNebula CMPInfo subclass
	"""

	@staticmethod
	def get_vm_list():
		server_url = "http://%s:%d/RPC2" % (ConfigONE.ONE_SERVER, ConfigONE.ONE_PORT)
		try:
			server = ServerProxy(server_url,allow_none=True,timeout=10)
			# To get only ONE_ID user's resources
			#vm_filter = -3
			# To get all
			vm_filter = -2
			(success, res_info, _) = server.one.vmpool.info(ConfigONE.ONE_ID, vm_filter, -1, -1, 3)
		except:
			logger.exception("Error getting the VM list")
			return []

		if success:
			res_vm = VM_POOL(res_info)
			res = []
			for vm in res_vm.VM:
				host = HostInfo(int(vm.HISTORY_RECORDS.HISTORY[0].HID), vm.HISTORY_RECORDS.HISTORY[0].HOSTNAME)
				new_vm = VirtualMachineInfo(int(vm.ID), host, int(vm.TEMPLATE.MEMORY) * 1024, vm)
				new_vm.user_id = vm.UID
				if vm.USER_TEMPLATE.MEM_TOTAL:
					# to make it work on all ONE versions
					real_memory = vm.TEMPLATE.REALMEMORY
					if not real_memory:
						real_memory = vm.REALMEMORY
					new_vm.set_memory_values(int(real_memory),
										int(vm.USER_TEMPLATE.MEM_TOTAL),
										int(vm.USER_TEMPLATE.MEM_FREE))
					if vm.USER_TEMPLATE.MIN_FREE_MEM:
						new_vm.min_free_mem = vm.USER_TEMPLATE.MIN_FREE_MEM
					if vm.USER_TEMPLATE.MEM_OVER:
						new_vm.mem_over_ratio = vm.USER_TEMPLATE.MEM_OVER
				
					# publish MEM properties to the VM user template to show the values to the user 
					OpenNebula._publish_mem_info(new_vm)

				res.append(new_vm)
				
			return res
		else:
			logger.error("Error getting the VM list: " + res_info)
			return []

	@staticmethod
	def _publish_mem_info(vm):
		"""
		Publish MIN_FREE_MEM and MEM_OVER properties to the VM user template to show the values to the user
		
		Args:
		- vm: VirtualMachineInfo with the VM info.

		Return: True if the information is published successfully or False otherwise 
		"""
		template = ""
		if not vm.min_free_mem:
			template += "MIN_FREE_MEM = %d\n" % Config.MIN_FREE_MEMORY
		if not vm.mem_over_ratio:
			template += "MEM_OVER = %.2f\n" % Config.MEM_OVER
		
		# if there is nothing to update return True
		if not template:
			return True 
		
		server_url = "http://%s:%d/RPC2" % (ConfigONE.ONE_SERVER, ConfigONE.ONE_PORT)
		try:
			server = ServerProxy(server_url,allow_none=True,timeout=10)
			(success, res_info, _) = server.one.vm.update(ConfigONE.ONE_ID, vm.id, template, 1)
			if not success:
				logger.error("Error updating the template to show the mem info to the VM ID: %s. %s." % (vm.id, res_info))
			return success
		except:
			logger.exception("Error updating the template to show the mem info to the VM ID: %s." % vm.id)
			return False
		
		return True
	
	@staticmethod
	def get_host_info(host_id):
		server_url = "http://%s:%d/RPC2" % (ConfigONE.ONE_SERVER, ConfigONE.ONE_PORT)
		try:
			server = ServerProxy(server_url,allow_none=True,timeout=10)
			(success, res_info, _) = server.one.host.info(ConfigONE.ONE_ID, host_id)
		except:
			logger.exception("Error getting the host info: " + host_id)
			return None
		
		if success:
			host_info = HOST(res_info)
			res_host = HostInfo(int(host_info.ID), host_info.NAME, host_info.STATE not in HOST.INVALID_STATES, host_info) 
			return res_host
		else:
			logger.error("Error getting the host info: " + res_info)
			return None
	
	@staticmethod
	def get_host_list():
		server_url = "http://%s:%d/RPC2" % (ConfigONE.ONE_SERVER, ConfigONE.ONE_PORT)
		try:
			server = ServerProxy(server_url,allow_none=True,timeout=10)
			(success, res_info, _) = server.one.hostpool.info(ConfigONE.ONE_ID)
		except:
			logger.exception("Error getting the host list")
			return None
		
		if success:
			res = []
			for host in HOST_POOL(res_info).HOST:
				new_host = HostInfo(host.ID, host.NAME, host.STATE not in HOST.INVALID_STATES, host)
				res.append(new_host)
			return res
		else:
			logger.error("Error getting the host list: " + res_info)
			return None
	
	@staticmethod
	def migrate(vm_id, host_id):
		server_url = "http://%s:%d/RPC2" % (ConfigONE.ONE_SERVER, ConfigONE.ONE_PORT)
		try:
			server = ServerProxy(server_url,allow_none=True,timeout=10)
			(success, res_info, _) = server.one.vm.migrate(ConfigONE.ONE_ID, vm_id, host_id, True, True)
		except:
			logger.exception("Error migrating the VM %d to the host %d" % (vm_id, host_id))
			return False
		
		if success:
			return True
		else:
			logger.error("Error migrating the VM %d to the host %d: %s" % (vm_id, host_id, res_info))
			return False

class MonitorONE(Monitor):
	"""
	Monitor for OpenNebula
	"""
	
	def __init__(self, cmpo = None):
		Monitor.__init__(self, OpenNebula())
	
	@staticmethod
	def host_has_memory_free(host_info,free_memory):
		"""
		Check if a node has enough free memory available
		"""
		if host_info:
			host_free_memory = host_info.raw.HOST_SHARE.FREE_MEM
			logger.debug("The host %s has %d KB of free memory." % (host_info.name, host_free_memory))
			if host_free_memory - free_memory > Config.HOST_MEM_MARGIN:
				return True
			else:
				return False
		else:
			return False

	def select_host_to_migrate(self, vm_info):
		"""
		Get the ID of the HOST to migrate. It selects the HOST with more free memory.
		If no node has enough memory or cpus to host the VM to migrate, return None.
		"""
		host_list = self.cmp.get_host_list()
		
		hosts_mem = {} 
		# Select the node with more memory free
		for host in host_list:
			hosts_mem[host] = host.raw.HOST_SHARE.FREE_MEM
		
		hosts_mem = sorted(hosts_mem.items(), key=lambda x: x[1], reverse = True)
		
		cpus = vm_info.raw.TEMPLATE.CPU
		if vm_info.total_memory:
			free_memory = vm_info.total_memory
		else:
			# If the monitored total memory is not available use the CMP original allocated one 
			free_memory = vm_info.allocated_memory

		powered = None
		while powered != False:
			for host, host_mem in hosts_mem:
				# ONE FREE_CPU is a Percentage
				host_free_cpus = host.raw.HOST_SHARE.FREE_CPU/100
				if host.active and host_mem > free_memory and host_free_cpus > cpus:
					return host
			
			# only try to poweron a host once
			if powered is None:
				# Let's try to power on a host
				powered = Monitor.power_on_host(free_memory, cpus)
			else:
				# otherwise continue
				powered = False
		
		return None

	@staticmethod
	def select_vm_to_migrate(req_vm_id, host_info, all_vms):
		"""
		Get the ID of the VM to migrate. It selects the VM with less memory avoiding to migrate VM with ID "req_vm_id"
		"""
		vms_mem = {} 
		for vm_id in host_info.raw.VMS.ID:
			for vm in all_vms:
				if int(vm_id) == vm.id and req_vm_id != vm.id:
					if vm.total_memory:
						vms_mem[vm] = vm.total_memory
					else:
						# If the monitored total memory is not available use the CMP original allocated one 
						vms_mem[vm] = vm.allocated_memory
		
		# if we want to get the biggest one set reverse=True
		vms_mem = sorted(vms_mem.items(), key=lambda x: x[1])
		
		return vms_mem[0][0]